infer/src/pulse/PulseExecutionDomain.ml

```diff
@@ -102,6 +108,170 @@ let pp fmt exec_state =
+          let termcond = Formula.extract_terminal_atoms cond in
+          let termcond2 = Formula.extract_terminal_terms cond in
+          let key = (cfgnode, termcond, pathcond, termcond2) in
+          let dl_ws = get_widenstate () in
+          match dl_ws with
+          | None ->
+              -1
+          | Some ws -> (
+              let prevstate = Stdlib.Hashtbl.find_opt ws key in
+              match prevstate with
+              | None ->
+                  let _ = add_widenstate key in
+                  record_pathcond tl (kl @ [key])
+              | Some _ -> (
+                match
+                  ( Formula.set_is_empty termcond
+                  , Formula.map_is_empty pathcond
+                  , Formula.termset_is_empty termcond2 )
+                with
+                | true, true, true ->
+                    idx
+                | _ ->
+                    let test = List.mem ~equal:cmp_four kl key in
+                    if test then -2 else idx ) ) )
+    in
+    let (repeated_wsidx : int) = if phys_equal same true then -1 else record_pathcond workset [] in
+    let create_infinite_state (hd : t) : t =
+      (* Only create infinite state from a non-error state that is not already an infinite state *)
+      match hd with
+      | ExceptionRaised astate ->
+          InfiniteProgram astate
+      | ContinueProgram astate ->
+          InfiniteProgram astate
+      | AbortProgram astate ->
+          AbortProgram astate
+      | ExitProgram astate ->
+          ExitProgram astate
+      | LatentAbortProgram a ->
+          LatentAbortProgram a
+      | LatentInvalidAccess a ->
+          LatentInvalidAccess a
+      | InfiniteProgram astate ->
+          InfiniteProgram astate
+      | LatentSpecializedTypeIssue astate ->
+          LatentSpecializedTypeIssue astate
+    in
+    let create_infinite_state_and_print (state_set : t list) (idx : int) (case : int) =
+      let _ = case in
+      Metadata.record_infinite_node cfgnode ;
+      let nth = List.nth state_set idx in
+      match nth with None -> ([], -1) | Some state -> ([create_infinite_state state], idx)
+    in
+    let nextempty = phys_equal (List.length next) 0 in
+    (* Identify which state in the post is repeated and generate a new infinite state for it *)
+    let rec is_repeated e lst : bool =
+      match lst with
+      | [] ->
+          false
+      | hd :: tl ->
+          if leq ~lhs:e ~rhs:hd then true else is_repeated e tl
+    in
+    let rec find_duplicate_state (lst : t list) (curi : int) (maxi : int) : int =
+      match lst with
+      | hd :: tl ->
+          let isrep = is_repeated hd tl in
+          if isrep && curi > maxi then find_duplicate_state tl (curi + 1) curi
+          else find_duplicate_state tl (curi + 1) maxi
+      | _ ->
+          maxi
+    in
+    (* we have a trivial lasso because prev = next - this should trigger an alert *)
+    if same && phys_equal nextempty false then
```

Comment:
use Bool.equal on bools, not phys_equal (could also just inline `List.is_empty`)
```suggestion
    if same && Bool.equal nextempty false then
```